src/service/dynamic_report/dynamic_report.py

```diff
@@ -0,0 +1,179 @@
+from app_names import COMPARISON_METRICS
+
+
+class DynamicReport:
+    def __init__(
+        self,
+        logger,
+        metric_report,
+        investment_report,
+        calendar_report,
+    ) -> None:
+        self.logger = logger
+        self.metric_report = metric_report
+        self.investment_report = investment_report
+        self.calendar_report = calendar_report
+
+    def __is_by_year_report(
+        self, metric: str, calendar_year: int, invest_year: int
+    ) -> bool:
+        return (calendar_year or (invest_year is not None)) and not metric
+
+    def __get_by_year_report_header(self) -> list:
+        return ["name", "sector", "vertical"] + COMPARISON_METRICS
+
+    def __comparison_report(self, header, company, peers) -> dict:
+        return {
+            "header": header,
+            "company_comparison_data": company,
+            "peers_comparison_data": peers,
+        }
+
+    def __get_metric_report(self, metric, access, username, **conditions) -> tuple:
+        self.investment_report.report.set_company_permissions(username)
+        years = self.metric_report.repository.get_years()
+        header = ["name"] + years
+        data = self.metric_report.get_by_metric_records(
+            metric, years, access, **conditions
+        )
+
+        return header, data
+
+    def __get_investment_report(
+        self, investment_year, access, username, **conditions
+    ) -> list:
+        self.investment_report.report.set_company_permissions(username)
+        data = self.investment_report.repository.get_base_metrics(
+            int(investment_year), **conditions
+        )
+
+        self.investment_report.add_calculated_metrics(data, access)
+        data = self.investment_report.report.filter_by_conditions(data, **conditions)
+
+        return data
+
+    def __get_dynamic_metrics(
```

Comment:
This function is too long, could you move the for implementation to another one?